packages/esm-patient-search-app/src/patient-search-icon/patient-search-icon.component.tsx

```diff
@@ -1,47 +1,163 @@
-import React, { useCallback, useEffect, useState } from 'react';
+import React, { useCallback, useState, useEffect, useReducer, useMemo } from 'react';
 import Search20 from '@carbon/icons-react/es/search/20';
+import Close20 from '@carbon/icons-react/es/close/20';
 import { HeaderGlobalAction } from 'carbon-components-react/es/components/UIShell';
 import styles from './patient-search-icon.component.scss';
 import PatientSearch from '../patient-search/patient-search.component';
+import Search from 'carbon-components-react/es/components/Search';
+import { useTranslation } from 'react-i18next';
+import { Button } from 'carbon-components-react';
+import debounce from 'lodash-es/debounce';
+import { useLayoutType } from '@openmrs/esm-framework';
+import { performPatientSearch } from '../patient-search/patient-search.resource';
+import isEmpty from 'lodash-es/isEmpty';
+import { SearchedPatient } from '../types';
 
 interface PatientSearchLaunchProps {}
 
+enum ActionTypes {
+  searching = 'searching',
+  resolved = 'resolved',
+  error = 'error',
+  idle = 'idle',
+}
+interface Searching {
+  type: ActionTypes.searching;
+  payload: Array<SearchedPatient>;
+}
+
+interface Error {
+  type: ActionTypes.error;
+  payload: Error;
+}
+interface Resolved {
+  type: ActionTypes.resolved;
+  payload: Array<SearchedPatient>;
+}
+
+interface Idle {
+  type: ActionTypes.idle;
+  payload: Array<SearchedPatient>;
+}
+
+type Action = Searching | Error | Resolved | Idle;
+
+interface PatientSearch {
+  status: 'searching' | 'resolved' | 'error' | 'idle';
+  searchResults: Array<SearchedPatient>;
+}
+
+function reducer(state: PatientSearch, action: Action): PatientSearch {
+  switch (action.type) {
+    case ActionTypes.resolved:
+      return { status: 'resolved', searchResults: action.payload };
+    case ActionTypes.searching:
+      return { status: 'searching', searchResults: action.payload };
+    case ActionTypes.error:
+      return { ...state, status: 'error' };
+    case ActionTypes.idle:
+      return { status: 'idle', searchResults: action.payload };
+  }
+}
+
+const searchTimeout = 300;
+const customRepresentation =
+  'custom:(patientId,uuid,identifiers,display,' +
+  'patientIdentifier:(uuid,identifier),' +
+  'person:(gender,age,birthdate,birthdateEstimated,personName,display),' +
+  'attributes:(value,attributeType:(name)))';
+
 const PatientSearchLaunch: React.FC<PatientSearchLaunchProps> = () => {
+  const { t } = useTranslation();
+  const layout = useLayoutType();
   const [open, setOpen] = useState<boolean>(false);
+  const [searchTerm, setSearchTerm] = useState<string>();
+  const initialState: PatientSearch = { status: 'idle', searchResults: [] };
+  const [{ searchResults, status }, dispatch] = useReducer(reducer, initialState);
 
-  const togglePatientSearch = useCallback(() => {
-    setOpen((prevState) => !prevState);
+  const performSearch = () => {
```

Comment:
The whole callback here screams for being used inside a side effect hook.

In the current state

- cancellation will never work
- if the component is unmounted during the resource loading we get state errors left and right